src/plugins/embeddable/public/react_embeddable_system/types.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import { SerializedPanelState } from '@kbn/presentation-containers';
+import { DefaultPresentationPanelApi } from '@kbn/presentation-panel-plugin/public/panel_component/types';
+import { PublishesUnsavedChanges, PublishingSubject } from '@kbn/presentation-publishing';
+import { ReactElement } from 'react';
+
+export type ReactEmbeddableRegistration<
+  ApiType extends DefaultEmbeddableApi = DefaultEmbeddableApi
+> = (ref: React.ForwardedRef<ApiType>) => ReactElement | null;
+
+/**
+ * The default embeddable API that all Embeddables must implement.
+ *
+ * Before adding anything to this interface, please be certain that it belongs in *every* embeddable.
+ */
+export type DefaultEmbeddableApi = DefaultPresentationPanelApi &
+  PublishesUnsavedChanges & {
+    serializeState: () => Promise<SerializedPanelState>;
+  };
+
+export type ReactEmbeddable<ApiType extends DefaultEmbeddableApi = DefaultEmbeddableApi> =
+  React.ForwardRefExoticComponent<React.RefAttributes<ApiType>>;
+
+export interface ReactEmbeddableFactory<
+  StateType extends unknown = unknown,
+  APIType extends DefaultEmbeddableApi = DefaultEmbeddableApi
+> {
+  getComponent: (initialState: StateType, maybeId?: string) => Promise<ReactEmbeddable<APIType>>;
+  deserializeState: (state: SerializedPanelState) => StateType;
+  latestVersion?: string;
+}
+
+export type StateTypeFromFactory<F extends ReactEmbeddableFactory<any>> =
+  F extends ReactEmbeddableFactory<infer S> ? S : never;
+
+/**
+ * State comparators
+ */
+export type EmbeddableComparatorFunction<StateType, KeyType extends keyof StateType> = (
+  last: StateType[KeyType] | undefined,
+  current: StateType[KeyType] | undefined,
+  lastState?: Partial<StateType>,
+  currentState?: Partial<StateType>
+) => boolean;
+
+export type EmbeddableComparatorDefinition<StateType, KeyType extends keyof StateType> = [
```

Comment:
Why is the definition stored in an array vs an object? Access parts of the defintion is cryptic with values like `comparators[key]?.[2]` instead of accessing a named key like `comparators[key]?.comparatorFunction`. I would recommend converting this type to an object with named keys unless there is a good reason for an array.

Reply:
Yes, the cryptic way to access the are not _ideal_.

This is somewhat mitigated by the fact that it's a tuple, not an array, meaning that Typescript can correctly infer the different types of each element.

The reason it's implemented this way is a tradeoff of more complicated **definition** code for the sake of less verbose **adoption** code. To illustrate, compare the example below:

## Object
``` typescript
const { unsavedChanges, resetUnsavedChanges } = useReactEmbeddableUnsavedChanges(
  uuid,
  markdownEmbeddableFactory,
  {
    firstName: {
      subject: firstName$,
      setter: setFirstName,
    },
    lastName: {
      subject: lastName$,
      setter: setLastName,
    },
    profession: {
      subject: profession$,
      setter: setProfession,
    },
    age: {
      subject: age$,
      setter: setAge,
      comparatorFunction: compareAges,
    },
    otherKey: {
      subject: otherKey$,
      setter: setOtherKey,
    },
  }
);
```

## Tuple
``` typescript
const { unsavedChanges, resetUnsavedChanges } = useReactEmbeddableUnsavedChanges(
  uuid,
  markdownEmbeddableFactory,
  {
    firstName: [firstName$, setFirstName],
    lastName: [lastName$, setLastName],
    profession: [profession$, setProfession],
    age: [age$, setAge, compareAges],
    otherKey: [otherKey$, setOtherKey],
  }
);
```

Consider that this is meant to be used as a hook inside an already fairly complicated component as well, that we anticipate some Embeddables to have _a lot_ of state keys, and that the tuple setup is very similar to how other react hooks work.

I've attempted to remove some of the confusion in 1f94fd4e53c by adding names like `const subject = comparator[0]`